File: packages/sdk-python/devbolt/hasher.py
"""Hash utilities for deterministic rollout distribution."""

import hashlib
from typing import Optional


class Hasher:
    """Deterministic hash function for rollout distribution."""

    DEFAULT_SEED = "devbolt"

    @staticmethod
    def get_bucket(flag_name: str, identifier: str, seed: Optional[str] = None) -> int:
        """Generate consistent hash bucket (0-99) for identifier.

        Args:
            flag_name: Name of the flag
            identifier: User identifier
            seed: Optional custom seed

        Returns:
            Bucket number (0-99)
        """
        hash_input = f"{seed or Hasher.DEFAULT_SEED}:{flag_name}:{identifier}"
        hash_digest = hashlib.sha256(hash_input.encode()).hexdigest()

        # Convert first 8 hex characters to number
        hash_number = int(hash_digest[:8], 16)

        # Return bucket 0-99
        return hash_number % 100

    @staticmethod
    def is_in_rollout(
        flag_name: str, identifier: str, percentage: float, seed: Optional[str] = None
    ) -> bool:
        """Check if identifier is in rollout percentage.

        Args:
            flag_name: Name of the flag
            identifier: User identifier
            percentage: Rollout percentage (0-100)
            seed: Optional custom seed

        Returns:
            True if in rollout, False otherwise
        """
        if percentage == 0:
            return False
        if percentage == 100:
            return True

        bucket = Hasher.get_bucket(flag_name, identifier, seed)
        return bucket < percentage

File: packages/core/src/hasher.ts
import { createHash } from "crypto";

/**
 * Deterministic hash function for rollout distribution
 */
export class Hasher {
  private static readonly DEFAULT_SEED = "devbolt";

  /**
   * Generate a consistent hash bucket (0-99) for a given identifier
   */
  static getBucket(
    flagName: string,
    identifier: string,
    seed?: string
  ): number {
    const hashInput = `${seed || this.DEFAULT_SEED}:${flagName}:${identifier}`;
    const hash = createHash("sha256").update(hashInput).digest("hex");

    // Convert first 8 hex characters to number
    const hashNumber = parseInt(hash.substring(0, 8), 16);

    // Return bucket 0-99
    return hashNumber % 100;
  }

  /**
   * Check if identifier is in rollout percentage
   */
  static isInRollout(
    flagName: string,
    identifier: string,
    percentage: number,
    seed?: string
  ): boolean {
    if (percentage === 0) return false;
    if (percentage === 100) return true;

    const bucket = this.getBucket(flagName, identifier, seed);
    return bucket < percentage;
  }
}
